storage/storage.go

```diff
@@ -31,22 +32,211 @@ func NewPaginationOptions(ps int32, contToken string) PaginationOptions {
 	}
 }
 
-// TupleIterator is an iterator for Tuples. It is closed by explicitly calling Stop() or by calling Next() until it
-// returns an TupleIteratorDone error.
-type TupleIterator interface {
-	Next() (*openfgapb.Tuple, error)
-	// Stop will release any resources held by the iterator. It must be safe to be called multiple times.
+type Iterator[T any] interface {
+	Next() (T, error)
 	Stop()
 }
 
 // ObjectIterator is an iterator for Objects (type + id). It is closed by explicitly calling Stop() or by calling Next() until it
-// returns an ObjectIteratorDone error.
-type ObjectIterator interface {
-	Next() (string, error)
-	// Stop will release any resources held by the iterator. It must be safe to be called multiple times.
-	Stop()
+// returns an IteratorDone error.
+type ObjectIterator = Iterator[*openfgapb.Object]
+
+// TupleIterator is an iterator for Tuples. It is closed by explicitly calling Stop() or by calling Next() until it
+// returns an IteratorDone error.
+type TupleIterator = Iterator[*openfgapb.Tuple]
+
+// TupleKeyIterator is an iterator for TupleKeys. It is closed by explicitly calling Stop() or by calling Next() until it
+// returns an IteratorDone error.
+type TupleKeyIterator = Iterator[*openfgapb.TupleKey]
+
+// UniqueObjectIterator takes two ObjectIterators and yields an ObjectIterator that returns only distinct objects with the
+// duplicates removed.
+func UniqueObjectIterator(iter1, iter2 ObjectIterator) (ObjectIterator, error) {
+	objectMap := map[string]*openfgapb.Object{}
+
+	for {
+		val, err := iter1.Next()
```

Comment:
This will drain the whole iterator even in cases when it's not needed. While ListObjects it's the only current consumer, would rather not have this blow up later on.

Reply:
That's necessary is it not? You cannot accumulate the unique objects without draining the two iterators in totality.

Since the `UniqueObjectIterator` is only consumed for ListObjects I'd be happy to move this as a utility/helper to the `commands/list_objects.go` source file, but I think it's ok here is it not? I can add doc to the godoc string that makes it more clear that this drains boths iterators, returning the unique objects between the two of them.

Reply:
> You cannot accumulate the unique objects without draining the two iterators in totality.

The iterator can keep a map of unique objects and on Next() you could check if the object you are about to return has been added to it. If not, you return it and add it to the map.

Reply:
You still have to iterate over all of the objects though. I'm confused how this helps the issue? You can't return prematurely from the iterator because you haven't seen the objects the iterator may yield until you've iterated over them.

What am I missing here?

Reply:
> You still have to iterate over all of the objects though. 

If the customer cancels the request midway, in your current code you would have unnecessarily iterated over everything

Reply:
That's the point of `Stop()` is it not? If the request is cancelled midway then the caller calling this would have called `iter1.Stop()` and this would return.

Reply:
`Stop` is never happening here. This will drain iter1 and iter2 the moment this third iterator is created. The iterators stop being lazy.

Reply:
Alright @elbuo8 @miparnisari

I refactored this to lazily yield unique objects from the two iterators. I think this addresses your concern. Let me know what you think 👍 ! Sorry for the misunderstanding. It makes a lot more sense now what you guys were trying to point out. I just wasn't processing it the same way you were.